Added mediaPlaybackRequiresUserAction to WebView

locly opened the pull request:
Just added a pass through to the `WebView` for `mediaPlaybackRequiresUserAction` and `setMediaPlaybackRequiresUserGesture` to allow auto-playing audio and video elements

Crash-- commented on Libraries/Components/WebView/WebView.ios.js:
```diff
@@ -231,9 +231,16 @@ var WebView = React.createClass({
      * **NOTE** : "In order for video to play inline, not only does this
      * property need to be set to true, but the video element in the HTML
      * document must also include the webkit-playsinline attribute."
-     * @platform ios
```
Don't remove this line. `allowsInlineMediaPlayback` is an iOS specific property.

mkonicek commented:
Hey @locly, how did you test this? Can you add some iOS and Android screenshot showing what this does and how it works?

locly commented:
We've been using a patched version of 0.20 with this in for a couple of weeks. At the high level it just allows the `autoplay` attribute of `audio` and `video` elements to be recognised, but also allows JavaScript in the webview to play audio on demand without being in a touch callback cycle.

I've added some example code below of how it works. Most of it is generating a test html page with an audio element and autoplay defined. Works the same for a video file too.

Not sure if needs a test case because it just exposes the `mediaPlaybackRequiresUserAction` and `setMediaPlaybackRequiresUserGesture` native properties of the webviews on each platform.

Added a screenshot for completeness, but the merge implements more of a behaviour change rather than a visual change.

```js
const React = require('react-native')
const { WebView } = React

module.exports = React.createClass({
  render: function() {
    const html = `
      <!DOCTYPE html>
      <html>
        <head>
          <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
          <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=0, minimum-scale=1.0, maximum-scale=1.0">
        </head>
        <body>
          <audio autoplay controls src="https://locly.com/api/card/981c88081c36601020d81d1dc1e93424/asset/audio.mp3"></audio>
        </body>
      </html>
    `
    return (
      <WebView
        style={{flex:1}}
        source={{ html: html, baseUrl: '' }}
        allowsInlineMediaPlayback={true}
        javaScriptEnabled={true}
        mediaPlaybackRequiresUserAction={false} />
    )
  }
});
```

![screen shot 2016-02-22 at 09 05 23](https://cloud.githubusercontent.com/assets/7373644/13213816/714c078c-d943-11e5-9ab3-10eac644a621.png)

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/webview/ReactWebViewManager.java:
```diff
@@ -277,6 +278,11 @@ public void setUserAgent(WebView view, @Nullable String userAgent) {
 
+  @ReactProp(name = "mediaPlaybackRequiresUserAction")
+  public void setMediaPlaybackRequiresUserAction(WebView view, boolean requires) {
+    view.getSettings().setMediaPlaybackRequiresUserGesture(requires);
```
Can `getSettings` return null? Should this code be checking for that?

Thomas101 replied:
I don't think `getSettings` does return null. `setDomStorageEnabled` does pretty much the same thing and doesn't check for example

mkonicek replied:
> I don't think getSettings does return null

Can you check the source code to be 100% sure? Also the behavior can change across Android versions. It's better to always check for `null` to be safe. 

See https://en.wikipedia.org/wiki/Defensive_programming

locly replied:
The [Android docs](http://developer.android.com/reference/android/webkit/WebView.html#getSettings()) don't state that it can ever return `null` and there are 4 other established instances in the same class where `view.getSettings()` is used without checking for `null` so I'd say it's pretty safe not to check in this instance.

mkonicek replied:
Ah now I see what you meant by those established instances. OK you've convinced me :)

mkonicek commented on Libraries/Components/WebView/WebView.ios.js:
```diff
@@ -233,6 +233,13 @@ var WebView = React.createClass({
+    /**
+     * Determines whether HTML5 audio & videos require the user to tap before they can
+     * start playing
+     * default value `false`
```
nit: ... start playing. The default value is `false`.

mkonicek commented on Libraries/Components/WebView/WebView.android.js:
```diff
@@ -137,6 +137,14 @@ var WebView = React.createClass({
+     * default value `false`
+     * @platform ios
+     */
+    mediaPlaybackRequiresUserAction: PropTypes.bool,
```
nit: ... start playing. The default value is `false`.

mkonicek commented:
Thanks for the explanation @locly. Minor nits, please check those and then let's merge this.

locly commented:
@mkonicek those look good. Agree with @Thomas101 no need to check for null -- other established code doesn't

mkonicek commented:
> Agree with @Thomas101 no need to check for null -- other established code doesn't

We should still check for null to be safe, see https://en.wikipedia.org/wiki/Defensive_programming

mkonicek commented:
@locly Please address the nits so that we can merge this.

> nit: ... start playing. The default value is false.

mkonicek commented on Libraries/Components/WebView/WebView.android.js:
```diff
@@ -137,6 +137,12 @@ var WebView = React.createClass({
+
+    /**
+     * Determines whether HTML5 audio & videos require the user to tap before they can
+     * start playing. The default value is `false`
```
nit: "The default value is `false`." is a sentence. Please add a period at the end. Sorry if I've missed that in the previous review.

mkonicek commented:
Just minor nits and let's shipit :+1:

locly commented:
Added!

mkonicek commented:
@facebook-github-bot shipit

mkonicek commented:
> I tried this but doesn't work :(

@lucasfeliciano How did you try this (provide code snippet) and what didn't work (provide error message, screenshots, anything to show what's broken).

mkonicek commented:
@facebook-github-bot shipit

lucasfeliciano commented:
@mkonicek sorry for the delay.

i made those changes in my local project to reflect this PR but the webview still need to receive the user input to play a youtube video for instance.

I even tried to inject a JS code that clicks on the play button of youtube but the native player opens and wait for the user input to play it.

locly commented:
@lucasfeliciano it sounds like the youtube app is picking up the url intent and loading in the native app. I'm guessing you're embedding the video using an `<iframe>` or using the YouTube JavaScript api?

If you use it was some self-hosted media in a `<audio>` or `<video>` element it should work but the YouTube api's are probably doing some sniffing to get it to open in their own app

lucasfeliciano commented:
@locly I'm passing the url below as `source.uri` to the webview.
If you open this in the browser it is HTML5 with no <iframe> on it.
 
`https://www.youtube.com/embed/ZZ5LpwO-An4?rel=0&autoplay=1&showinfo=0&controls=0`

locly commented:
The link you're passing there is to a YouTube webpage that shows the video full screen. It's then running some scripts to see if it can open up in the native app. This fix is more about providing support for the autoplay attribute in html5 audio and video elements (https://developer.mozilla.org/en/docs/Web/HTML/Element/video).

You need to get a hold of the actual video file and embed it in your own page. For example...

```html
<html>
...
<body>
<video controls autoplay src="http://download.blender.org/peach/bigbuckbunny_movies/BigBuckBunny_320x180.mp4"></video>
</body>
</html>
```
